jaxrs-spec/src/main/asciidoc/chapters/resources/_declaring_method_capabilities.adoc

```diff
@@ -126,3 +129,58 @@ the media type `application/xml` given its higher qs-value. Note that
 qs-values, just like q-values, are relative and as such are only
 comparable to other qs-values within the same `@Produces` annotation
 instance. For more information see <<determine_response_type>>.
+
+[[consuming_multipart_formdata]]
+==== Special handling for consuming multipart/form-data media types
+
+The `multipart/form-data` media type enables requests to send multiple
+entities (parts) as a single entity. Each part contains it's own set of
+headers, media type and content. There are a few different ways to
+consume these requests in a resource method. First, a resource method
+may contain an entity parameter of type, `List<Part>`. For example:
+
+[source,java]
+----
+@Path("widgets3")
+public class WidgetsResource3 {
+
+    @POST
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response postWidget(List<Part> parts) {
+        for (Part part : parts) {
+            String name = part.getName();
+            Optional<String> fileName = part.getFileName();
+            InputStream is = part.getEntityStream();
+            MultivaluedMap<String, String> partHeaders = part.getHeaders();
+            MediaType mediaType = part.getMediaType();
+            doSomethingWithPart(name, fileName, is, partHeaders, mediaType);
+        }
+        return Response.ok().build();
+    }
+
+}
+----
+
+Another approach is to use `@FormParam` parameters where the value in the
+annotation corresponds to the name of the part. The parameter type may be
+a `jakarta.ws.rs.ext.Part`, a `java.io.InputStream` or a `String`. Here is
+an example:
+
+[source,java]
+----
+@Path("widgets4")
+public class WidgetsResource4 {
+
+    @POST
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response postWidget(@FormParam("part1Name") String part1, 
+                               @FormParam("part2Name") InputStream part2,
+                               @FormParam("part3Name") Part part3) {...}
+}
+----
+
+The only way to access the headers for a particular part is to use the `Part`
+type. The `InputStream` and `String` parameter types will only provide the
+content of the part. Note that parts of a multipart entity can be quite large,
+so care should be taken when using `String` parameter types as that will load
+the entire content of the part into the Java heap.
```

Comment:
>What is the character encoding used for the String case?

My interpretation of the RFC is that the encoding should be UTF-8 unless specified otherwise.  The RFC says that the encoding/charset can be specified per-part via the part's `Content-Type` header.  It also mentions a "hidden" part with the reserved name of "\_charset\_" and whose body would specify the charset for all other parts. 

I think I like the per-part approach better - and it is easy enough to specify with the APIs in this PR.

The "hidden" _charset_ part could also work with the current APIs, but would be a little clunky... maybe we want to create a quick helper method to generate a default charset part? Wdyt?

>  Should we support byte[]?

🤷  maybe??  I was hesitant to support `String` because it forces the content to be loaded into the heap - and that could be a vulnerability without proper safeguards.  I ended up pulling in `String` because RESTEasy supports it (I think CXF might also), so I figured users must want that - maybe for secured environment where they know that that parts are small enough to handle (btw, both RESTEasy and CXF has config settings to limit the size of these attachments which should reduce the vulnerability).

If we support `byte[]` do you think we should drop support for `String`? That would put the encoding burden on the user - and they wouldn't be able to read the part's headers to see if it had a non-default encoding...  Wdyt?

Reply:
Yes, on UTF-8 and charset in `Content-Type`. The hidden part option seems like an _ugly hack_ to me. You're right about the issue with `charset` and `byte[]`, but it also applies to the `InputStream` alternative. It's simple to convert an `InputStream` to a `byte[]`, so the latter may not be really needed.